pageserver/src/tenant.rs

```diff
@@ -1756,21 +1760,23 @@ impl Tenant {
         let (state, mut rx) = watch::channel(state);
 
         tokio::spawn(async move {
-            let current_state = *rx.borrow_and_update();
+            let mut current_state = format!("{}", &*rx.borrow_and_update());
```

Comment:
I think there was a bug here, previously we didnt update current_state value, so on state change we've always decremented the value for initial state, not for the previous one

Reply:
why is formatted to a str instead of as_str for metric purposes?

Reply:
Unfortunately I cannot keep reference to borrowed value across loop iterations (and if I could the value would change underneath so current would be equal to new). 

Additionally with borrowed value (Ref returned from borrow_and_update) the future becomes !Send.

So replaced as_str with owned string.

The patch if you want to play with it:

```
diff --git a/pageserver/src/tenant.rs b/pageserver/src/tenant.rs
index 17f6f4ef..4feb3ce9 100644
--- a/pageserver/src/tenant.rs
+++ b/pageserver/src/tenant.rs
@@ -1515,7 +1515,7 @@ impl Tenant {
                         timeline.set_state(TimelineState::Stopping);
                     }
                 }
-                TenantState::Broken{..} => {
+                TenantState::Broken { .. } => {
                     info!("Cannot set tenant to Stopping state, it is already in Broken state due to: {:?}", current_state);
                 }
                 TenantState::Stopping => {
@@ -1760,20 +1760,20 @@ impl Tenant {
         let (state, mut rx) = watch::channel(state);
 
         tokio::spawn(async move {
-            let mut current_state = format!("{}", &*rx.borrow_and_update());
+            let mut current_state = rx.borrow_and_update();
             let tid = tenant_id.to_string();
             TENANT_STATE_METRIC
-                .with_label_values(&[&tid, &current_state])
+                .with_label_values(&[&tid, current_state.as_ref()])
                 .inc();
             loop {
                 match rx.changed().await {
                     Ok(()) => {
-                        let new_state = format!("{}", &*rx.borrow_and_update());
+                        let new_state = rx.borrow_and_update();
                         TENANT_STATE_METRIC
-                            .with_label_values(&[&tid, &current_state])
+                            .with_label_values(&[&tid, current_state.as_ref()])
                             .dec();
                         TENANT_STATE_METRIC
-                            .with_label_values(&[&tid, &new_state])
+                            .with_label_values(&[&tid, new_state.as_ref()])
                             .inc();
 
                         current_state = new_state;
```

Reply:
Ok, perhaps the generated as_str is no longer returning `&'static str`? Feels a bit weird if it wasn't a static str. Could it be an issue that the TenantState is being kept a reference to and not just prev-state label str to decrement? I probably wont have time to try this out. I think the patch would keep a reference to the TenantState instead of it's static str.

My second worry was, as we no longer have a clear as_str impl, that it might contain the stacktrace and reason, which it hopefully does not contain, but I cant remember seeing an Display impl either.

But yeah, probably fine regardless.. Don't want to block this.

Reply:
Gotcha, the idea is to keep the 'static str reference. 

There is separate [IntoStaticStr](https://docs.rs/strum_macros/0.24.3/strum_macros/derive.IntoStaticStr.html). I've updated the code to use it and it worked

Reply:
Ok then I am now very confused what does `AsStr` derive do, because I would had thought it would be this `IntoStaticStr`...